c7n/resources/ami.py

```diff
@@ -295,6 +295,40 @@ def process_image(self, client, image):
             LaunchPermission={'Remove': remove},
             OperationType='remove')
 
+@AMI.action_registry.register('cancel-launch-permissions')
+class CancelLaunchPermissions(BaseAction):
+    """Action to cancel the shared AMI to your account
+
+    AWS account and you no longer want it shared with your account,
+    you can remove your account from the AMI's launch permissions
+
+    :example:
+
+    .. code-block:: yaml
+
+            policies:
+              - name: ami-cancel-share-to-me-old
+                resource: ami
+                query:
+                  - ExecutableUsers: [self]
+                  - Owners: []
+                actions:
+                  - type: cancel-launch-permissions
+
+    """
+    schema = type_schema('cancel-launch-permissions', value={'type': 'boolean'})
+
+    permissions = ('ec2:CancelImageLaunchPermission',)
+
+    def process(self, images):
+        client = local_session(self.manager.session_factory).client('ec2')
+        for i in images:
+            self.process_image(client, i)
+
+    def process_image(self, client, image):
+        client.cancel_image_launch_permission(
+            ImageId=image['ImageId'],
+            DryRun=self.data.get('value', True))
```

Comment:
To @kapilt 's point, this won't remove permissions that come from org/OU-level shares. From some local testing, it looks like you get the same error (`InvalidAMIID.NotFound`) in all of these cases:

- The AMI doesn't exist
- The AMI exists, but is not shared with you
- The AMI exists, and is shared with you through org/OU-level permissions

I don't think we want to bomb a policy run on that exception, but I don't think we want to completely ignore it either. Perhaps a warning just to show what's going on in that case is the best we can do. Something like:

```suggestion
        try:
            client.cancel_image_launch_permission(
                ImageId=image['ImageId'],
                DryRun=self.data.get('dry_run', True),
            )
        except ClientError as err:
            if err.response["Error"]["Code"] != "InvalidAMIID.NotFound":
                raise
            self.log.warning(err.response["Error"]["Message"])
```

Reply:
Thanks @ajkerrigan @kapilt  for your reviews and feedbacks, also once again thanks for @ajkerrigan applied the try exception.

Reply:
good call, re invalid id, we should handle that, got down the rabbit hole of remembering that boto ec2 doesn't have separate client exception classes for some reason.